### PR TITLE
Add new `Style/RedundantHeredocDelimiterQuotes` cop

### PR DESCRIPTION
**Pull request by koic**
Follow up https://github.com/rubocop/rubocop/pull/11522#discussion_r1091525976.

This cop checks for redundant heredoc delimiter quotes.

```ruby
# bad
do_something(<<~'EOS')
  no string interpolation style text
EOS

# good
do_something(<<~EOS)
  no string interpolation style text
EOS
```

-----------------

Before submitting the PR make sure the following are checked:

* [x] The PR relates to *only* one subject with a clear title and description in grammatically correct, complete sentences.
* [x] Wrote [good commit messages][1].
* [ ] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Ran `bundle exec rake default`. It executes all tests and runs RuboCop on its own code.
* [x] Added an entry (file) to the [changelog folder](https://github.com/rubocop/rubocop/blob/master/changelog/) named `{change_type}_{change_description}.md` if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format) for details.

[1]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

File: lib/rubocop/cop/style/redundant_heredoc_delimiter_quotes.rb [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+      # Checks for redundant heredoc delimiter quotes.
+      #
+      # @example
+      #
+      #   # bad
+      #   do_something(<<~'EOS')
+      #     no string interpolation style text
+      #   EOS
+      #
+      #   # good
```
[the remark this answers is not in the document]
Comment (koic) [reply lib/rubocop/cop/style/redundant_heredoc_delimiter_quotes.rb:15]: Certainly! I added some examples for that.

**Comment by bbatsov** [comment]
Pretty useful cop! Thanks!

File: lib/rubocop/cop/style/redundant_heredoc_delimiter_quotes.rb [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+      #   do_something(<<~'EOS')
+      #     escaped character\.
+      #   EOS
+      #
+      class RedundantHeredocDelimiterQuotes < Base
+        include Heredoc
+        extend AutoCorrector
+
+        MSG = 'Remove the redundant heredoc delimiter quotes, use `%<replacement>s` instead.'
+        STRING_INTERPOLATION_OR_ESCAPED_CHARACTER_PATTERN = /#(\{|@|\$)|\\/.freeze
```
Comment (splattael) [inline lib/rubocop/cop/style/redundant_heredoc_delimiter_quotes.rb:33]: @koic https://github.com/rubocop/rubocop/pull/11543 adds a missing spec to test escapes.

**Comment by javierjulio** [comment]
@koic thanks! Since we picked up this release through Dependabot, I check the docs to find more info but I didn't see the rule there. I realize now that 1.45 isn't published on docs.rubocop.org where this was included. Only 1.44 is available.

**Comment by koic** [comment]
@javierjulio The 1.45 doc is published. Thank you!
https://docs.rubocop.org/rubocop/1.45/index.html

**Comment by bbatsov** [comment]
My bad. As the process is manual and I constantly switch between computers from time to time I forget to do it when I release from the "wrong" computer (the one where I haven't setup Antora locally yet).

**Comment by bbatsov** [comment]
On the bright side - this finally motivated me to setup Node.js and Antora on my laptop as well. 😄

**Pull request merged**